src/storage/tracked.rs

```diff
@@ -0,0 +1,413 @@
+use std::iter::Enumerate;
+use std::marker::PhantomData;
+use std::ops::{Deref, DerefMut};
+use std::slice;
+
+use hibitset::{BitSet, BitSetLike};
+
+use {Component, DenseVecStorage, Index, MaskedStorage, Storage, UnprotectedStorage};
+
+/// This enum represents the different changes the `TrackedStorage`
+/// can detect.
+///
+/// ## Resets
+///
+/// You most likely want to reset the `TrackedStorage` every frame.
+/// By doing that, all change events are cleared.
+///
+/// ## Maintain
+///
+/// By maintaining the `FlaggedStorage`, the inner storage and the cache
+/// will be searched for unequal elements, generating `Modified` events for those.
+///
+/// ## Frames
+///
+/// In the following, the term "frame" will be used to describe
+/// the timeframe between a reset and an eventual maintain.
+/// (eventual maintain because you may not be interested in `Modified` events)
+///
+/// ## Which event will be generated?
+///
+/// This section explains some edge cases.
+///
+/// ### There are no insertions, no removals, no maintained changes
+///
+/// No event (`None`)
+///
+/// ### There has been one insertion but also one removal in this frame
+///
+/// Because the change has been reverted, there's no event (`None`)
+///
+/// ### There has been a removal and an insertion in this frame
+///
+/// This is the only case where a `Modified` event will be generated
+/// without a call to `maintain`.
+///
+/// ### A component has been inserted and modified later in this frame (maintained)
+///
+/// The insertion has the bigger priority (`Inserted`).
+///
+/// ### A component has been modified and then removed
+///
+/// The removal has priority over the modification (`Removed`).
+#[derive(Copy, Clone, Debug, PartialEq, PartialOrd)]
+pub enum Change {
+    /// `None`s are filtered out and the user should never need to deal with it.
+    #[doc(hidden)]
+    None,
+    /// A component has been inserted.
+    Inserted,
+    /// The component has been modified.
+    Modified,
+    /// The component has been removed.
+    Removed,
+}
+
+impl Change {
+    fn add(&mut self, change: Change) {
+        use Change::*;
+
+        let new = match (*self, change) {
+            (None, new) => new,
+            (Inserted, Modified) => Inserted,
+            (Inserted, Removed) => None,
+            (Modified, Modified) => Modified,
+            (Modified, Removed) => Removed,
+            (Removed, Inserted) => Modified,
+            (old, new) => panic!("Didn't expect change from {:?} to {:?}", old, new),
+        };
+
+        *self = new;
+    }
+}
+
+/// Events iterator for all events between the last reset and the last,
+/// eventual maintain. See `Change` docs for more explanations.
```

Comment:
"last, eventual maintain" ?

Reply:
It's explained in `Change` docs. If the component doesn't implement `PartialEq`, you can also use the `TrackedStorage`, you just don't get any modification events but just insertions and removals.

Reply:
If it's the grammar that's broken please suggest a better wording.